File: lib/lunarml/mlinit.js
"use strict";
//BEGIN _list
function _list(a) {
    let x = null;
    for (let i = a.length - 1; i >= 0; --i) {
        x = [a[i], x];
    }
    return x;
}
//END
//BEGIN _Match_tag
function _Match_tag() {}
_Match_tag.prototype.__isMLExn = true;
_Match_tag.prototype.name = "Match";
//END
//BEGIN _Match: _Match_tag
const _Match = new _Match_tag();
//END
//BEGIN _isMatch: _Match_tag
function _isMatch(e) { return e instanceof _Match_tag; }
//END
//BEGIN _Bind_tag
function _Bind_tag() {}
_Bind_tag.prototype.__isMLExn = true;
_Bind_tag.prototype.name = "Bind";
//END
//BEGIN _Bind: _Bind_tag
const _Bind = new _Bind_tag();
//END
//BEGIN _isBind: _Bind_tag
function _isBind(e) { return e instanceof _Bind_tag; }
//END
//BEGIN _Div_tag
function _Div_tag() {}
_Div_tag.prototype.__isMLExn = true;
_Div_tag.prototype.name = "Div";
//END
//BEGIN _Div: _Div_tag
const _Div = new _Div_tag();
//END
//BEGIN _isDiv: _Div_tag
function _isDiv(e) { return e instanceof _Div_tag; }
//END
//BEGIN _Overflow_tag
function _Overflow_tag() {}
_Overflow_tag.prototype.__isMLExn = true;
_Overflow_tag.prototype.name = "Overflow";
//END
//BEGIN _Overflow: _Overflow_tag
const _Overflow = new _Overflow_tag();
//END
//BEGIN _isOverflow: _Overflow_tag
function _isOverflow(e) { return e instanceof _Overflow_tag; }
//END
//BEGIN _Size_tag
function _Size_tag() {}
_Size_tag.prototype.__isMLExn = true;
_Size_tag.prototype.name = "Size";
//END
//BEGIN _Size: _Size_tag
const _Size = new _Size_tag();
//END
//BEGIN _isSize: _Size_tag
function _isSize(e) { return e instanceof _Size_tag; }
//END
//BEGIN _Subscript_tag
function _Subscript_tag() {}
_Subscript_tag.prototype.__isMLExn = true;
_Subscript_tag.prototype.name = "Subscript";
//END
//BEGIN _Subscript: _Subscript_tag
const _Subscript = new _Subscript_tag();
//END
//BEGIN _isSubscript: _Subscript_tag
function _isSubscript(e) { return e instanceof _Subscript_tag; }
//END
//BEGIN _Fail_tag
function _Fail_tag(payload) { this.payload = payload; }
_Fail_tag.prototype.__isMLExn = true;
_Fail_tag.prototype.name = "Fail";
//END
//BEGIN _Fail: _Fail_tag
function _Fail(payload) { return new _Fail_tag(payload); }
//END
//BEGIN _isFail: _Fail_tag
function _isFail(e) { return e instanceof _Fail_tag; }
//END
//BEGIN _Fail_payload
function _Fail_payload(e) { return e.payload; }
//END
//BEGIN _id
function _id(x) { return x; }
//END
//BEGIN _isError
function _isError(e) { return !("__isMLExn" in e); }
//END
//BEGIN _mkFn2
function _mkFn2(f) {
  return (a, b) => f([a, b]);
}
//END
//BEGIN _mkFn3
function _mkFn3(f) {
  return (a, b, c) => f([a, b, c]);
}
//END
//BEGIN _String_EQUAL
function _String_EQUAL(s, t) {
    if (s === t) { return true; }
    const n = s.length;
    if (n !== t.length) { return false; }
    for (let i = 0; i < n; ++i) {
        if (s[i] !== t[i]) {
            return false;
        }
    }
    return true;
}
//END
//BEGIN MIN_INT32
const MIN_INT32 = -0x80000000;
//END
//BEGIN MAX_INT32
const MAX_INT32 = 0x7fffffff;
//END
//BEGIN _Int32_abs: MIN_INT32 _Overflow
function _Int32_abs(x) {
    if (x < 0) {
        if (x === MIN_INT32) {
            throw _Overflow;
        }
        return -x;
    } else {
        return x;
    }
}
//END
//BEGIN _Int32_negate: MIN_INT32 _Overflow
function _Int32_negate(x) {
    if (x === MIN_INT32) {
        throw _Overflow;
    }
    return (-x)|0;
}
//END
//BEGIN _Int32_add: MIN_INT32 MAX_INT32 _Overflow
function _Int32_add(x, y) {
    const z = x + y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z;
}
//END
//BEGIN _Int32_sub: MIN_INT32 MAX_INT32 _Overflow
function _Int32_sub(x, y) {
    const z = x - y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z;
}
//END
//BEGIN _Int32_mul: MIN_INT32 MAX_INT32 _Overflow
function _Int32_mul(x, y) {
    const z = x * y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z|0;
}
//END
//BEGIN _Int32_div: _Div MIN_INT32 _Overflow
function _Int32_div(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT32 && y === -1) {
        throw _Overflow;
    }
    return Math.floor(x / y)|0;
}
//END
//BEGIN _Int32_mod: _Div
function _Int32_mod(x, y) {
    if (y === 0) {
        throw _Div;
    }
    return x - Math.floor(x / y) * y;
}
//END
//BEGIN _Int32_quot: MIN_INT32 _Overflow
function _Int32_quot(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT32 && y === -1) {
        throw _Overflow;
    }
    return (x / y)|0;
}
//END
//BEGIN MIN_INT54
const MIN_INT54 = -0x20000000000000;
//END
//BEGIN MAX_INT54
const MAX_INT54 = 0x1fffffffffffff;
//END
//BEGIN _Int54_abs: MIN_INT54 _Overflow
function _Int54_abs(x) {
    if (x < 0) {
        if (x === MIN_INT54) {
            throw _Overflow;
        }
        return -x;
    } else {
        return x;
    }
}
//END
//BEGIN _Int54_negate: MIN_INT54 _Overflow
function _Int54_negate(x) {
    if (x === MIN_INT54) {
        throw _Overflow;
    }
    return 0 - x; // Avoid -0
}
//END
//BEGIN _Int54_add: MIN_INT54 MAX_INT54 _Overflow
function _Int54_add(x, y) {
    const z = x + y;
    if ((MIN_INT54 < z && z <= MAX_INT54) || (z === MIN_INT54 && (x & 1) === (y & 1))) {
        return z;
    } else {
        throw _Overflow;
    }
}
//END
//BEGIN _Int54_sub: MIN_INT54 MAX_INT54 _Overflow
function _Int54_sub(x, y) {
    const z = x - y;
    if ((MIN_INT54 < z && z <= MAX_INT54) || (z === MIN_INT54 && (x & 1) === (y & 1))) {
        return z;
    } else {
        throw _Overflow;
    }
}
//END
//BEGIN _Int54_mul: MIN_INT54 MAX_INT54 _Overflow
function _Int54_mul(x, y) {
    const z = 0 + x * y;
    if ((MIN_INT54 < z && z <= MAX_INT54) || (z === MIN_INT54 && ((x & 1) === 0 || (y & 1) === 0))) {
        return z;
    } else {
        throw _Overflow;
    }
}
//END
//BEGIN _Int54_div: _Div MIN_INT54 _Overflow
function _Int54_div(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT54 && y === -1) {
        throw _Overflow;
    } else {
        return 0 + Math.floor(x / y);
    }
}
//END
//BEGIN _Int54_mod: _Div
function _Int54_mod(x, y) {
    if (y === 0) {
        throw _Div;
    } else {
        const r = 0 + x % y;
        if (r === 0 || x * y >= 0) {
            return r;
        } else {
            return r + y;
        }
    }
}
//END
//BEGIN _Int54_quot: _Div MIN_INT54 _Overflow
function _Int54_quot(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT54 && y === -1) {
        throw _Overflow;
    } else {
        return 0 + Math.trunc(x / y);
    }
}
//END
//BEGIN _Int54_rem: _Div
function _Int54_rem(x, y) {
    if (y === 0) {
        throw _Div;
    } else {
        return 0 + x % y;
    }
}
//END
//BEGIN Math_abs
const Math_abs = Math.abs;
//END
//BEGIN Math_imul
const Math_imul = Math.imul;
//END
//BEGIN _encodeUtf8
function _encodeUtf8(s) {
    const encoder = new TextEncoder();
    return encoder.encode(s);
}
//END
//BEGIN _decodeUtf8
function _decodeUtf8(s) {
    const decoder = new TextDecoder();
    return decoder.decode(s);
}
//END
//BEGIN _exnName: _encodeUtf8
function _exnName(e) { return _encodeUtf8(e.name); }
//END
//BEGIN _String_LT
function _String_LT(a, b) {
    let i = 0;
    const m = a.length, n = b.length;
    while (i < m && i < n) {
        const x = a[i], y = b[i];
        if (x < y) {
            return true;
        } else if (x > y) {
            return false;
        }
        ++i;
    }
    return m < n;
}
//END
//BEGIN _String_append
function _String_append(a, b) {
    if (a.length === 0) { return b; }
    if (b.length === 0) { return a; }
    const c = new Uint8Array(a.length + b.length);
    c.set(a);
    c.set(b, a.length);
    return c;
}
//END
//BEGIN _String_concat
function _String_concat(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const s = xs0[0];
        n += s.length;
        xs0 = xs0[1];
    }
    const a = new Uint8Array(n);
    let m = 0;
    while (xs !== null) {
        const s = xs[0];
        a.set(s, m);
        m += s.length;
        xs = xs[1];
    }
    return a;
}
//END
//BEGIN _String_concatWith
function _String_concatWith(sep, xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const s = xs0[0];
        n += s.length;
        xs0 = xs0[1];
        if (xs0 !== null) {
            n += sep.length;
        }
    }
    const a = new Uint8Array(n);
    let m = 0;
    while (xs !== null) {
        const s = xs[0];
        a.set(s, m);
        m += s.length;
        xs = xs[1];
        if (xs !== null) {
            a.set(sep, m);
            m += sep.length;
        }
    }
    return a;
}
//END
//BEGIN _String_implode
function _String_implode(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        ++n;
        xs0 = xs0[1];
    }
    const a = new Uint8Array(n);
    let i = 0;
    while (xs !== null) {
        a[i] = xs[0];
        xs = xs[1];
        ++i;
    }
    return a;
}
//END
//BEGIN _String_translate
function _String_translate(f, s) {
    const m = s.length;
    const a = new Array(m);
    let n = 0;
    for (let i = 0; i < m; ++i) {
        const t = f(s[i]);
        a[i] = t;
        n += t.length;
    }
    const r = new Uint8Array(n);
    let l = 0;
    for (let i = 0; i < m; ++i) {
        r.set(a[i], l);
        l += a[i].length;
    }
    return r;
}
//END
//BEGIN _String32_append
function _String32_append(a, b) {
    if (a.length === 0) { return b; }
    if (b.length === 0) { return a; }
    const c = new Int32Array(a.length + b.length);
    c.set(a);
    c.set(b, a.length);
    return c;
}
//END
//BEGIN _String32_concat
function _String32_concat(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const s = xs0[0];
        n += s.length;
        xs0 = xs0[1];
    }
    const a = new Int32Array(n);
    let m = 0;
    while (xs !== null) {
        const s = xs[0];
        a.set(s, m);
        m += s.length;
        xs = xs[1];
    }
    return a;
}
//END
//BEGIN _String32_concatWith
function _String32_concatWith(sep, xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const s = xs0[0];
        n += s.length;
        xs0 = xs0[1];
        if (xs0 !== null) {
            n += sep.length;
        }
    }
    const a = new Int32Array(n);
    let m = 0;
    while (xs !== null) {
        const s = xs[0];
        a.set(s, m);
        m += s.length;
        xs = xs[1];
        if (xs !== null) {
            a.set(sep, m);
            m += sep.length;
        }
    }
    return a;
}
//END
//BEGIN _String32_implode
function _String32_implode(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        ++n;
        xs0 = xs0[1];
    }
    const a = new Int32Array(n);
    let i = 0;
    while (xs !== null) {
        a[i] = xs[0];
        xs = xs[1];
        ++i;
    }
    return a;
}
//END
//BEGIN _String32_translate
function _String32_translate(f, s) {
    const m = s.length;
    const a = new Array(m);
    let n = 0;
    for (let i = 0; i < m; ++i) {
        const t = f(s[i]);
        a[i] = t;
        n += t.length;
    }
    const r = new Int32Array(n);
    let l = 0;
    for (let i = 0; i < m; ++i) {
        r.set(a[i], l);
        l += a[i].length;
    }
    return r;
}
//END
//BEGIN _UTF16_LT
function _UTF16_LT(a, b) {
    let i = 0;
    const m = a.length, n = b.length;
    while (i < m && i < n) {
        const x = a.codePointAt(i), y = b.codePointAt(i);
        if (x !== y) {
            return x < y;
        }
        i += 1 + (x >= 0x10000);
    }
    return m < n;
}
//END
//BEGIN _UTF16_size
function _UTF16_size(a) {
    const n = a.length;
    let l = 0;
    for (let i = 0; i < n; ) {
        const x = a.codePointAt(i);
        i += 1 + (x >= 0x10000);
        l += 1;
    }
    return l;
}
//END
//BEGIN _UTF16_isWellFormed
const _UTF16_isWellFormed =
    typeof "".isWellFormed === "function"
    ? (s => s.isWellFormed())
    : (s => s.search(/[\uD800-\uDFFF]/u) === -1);
//END
//BEGIN _UTF16_toWellFormed
const _UTF16_toWellFormed =
    typeof "".toWellFormed === "function"
    ? (s => s.toWellFormed())
    : (s => s.replace(/[\uD800-\uDFFF]/gu, "\uFFFD"));
//END
//BEGIN _Array_array: _Size
function _Array_array(n, init) {
    if (n < 0 || n > 0xffffffff) {
        throw _Size;
    }
    const a = new Array(n);
    a.fill(init);
    return a;
}
//END
//BEGIN _VectorOrArray_fromList
function _VectorOrArray_fromList(xs) {
    const a = [];
    while (xs !== null) {
        a.push(xs[0]);
        xs = xs[1];
    }
    return a;
}
//END
//BEGIN _Vector_unsafeFromListRevN: _Size
function _Vector_unsafeFromListRevN(n, xs) {
    if (n > 0xffffffff) {
        throw _Size;
    }
    const a = new Array(n);
    let i = n - 1;
    while (xs !== null) {
        a[i] = xs[0];
        xs = xs[1];
        --i;
    }
    // i should be -1 here
    return a;
}
//END
//BEGIN _VectorOrArray_tabulate: _Size
function _VectorOrArray_tabulate(n, f) {
    if (n < 0 || n > 0xffffffff) {
        throw _Size;
    }
    const a = new Array(n);
    for (let i = 0; i < n; ++i) {
        a[i] = f(i);
    }
    return a;
}
//END
//BEGIN _Vector_concat: _Size
function _Vector_concat(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const v = xs0[0];
        n += v.length;
        xs0 = xs0[1];
    }
    if (n > 0xffffffff) {
        throw _Size;
    }
    const a = new Array(n);
    let m = 0;
    while (xs !== null) {
        const v = xs[0];
        for (let i = 0; i < v.length; ++i) {
            a[m++] = v[i];
        }
        xs = xs[1];
    }
    return a;
}
//END
//BEGIN _ThenableWrapper
function _ThenableWrapper(wrapped) {
    this.payload = wrapped;
}
//END
//BEGIN _wrapThenable: _ThenableWrapper
function _wrapThenable(x) {
    if (x != null && (typeof x === "object" || typeof x === "function")) {
        try {
            if (typeof x.then === "function") {
                return new _ThenableWrapper(x);
            }
        } catch {
        }
    }
    return x;
}
//END
//BEGIN _unwrapThenable: _ThenableWrapper
function _unwrapThenable(x) {
    return x instanceof _ThenableWrapper ? x.payload : x;
}
//END
/*
type Trampoline<Result> = [true, Result] | exists Args. [false, MLFunction<Result, Args>, Args]
interface MLFunction<Result, Args> {
    (...args: Args): Result;
    _MLTAIL_?: (...args: Args) => Trampoline<Result>;
}
*/
//BEGIN _wrap
function _wrap(f) {
    const F = function() {
        let result = f.apply(undefined, arguments); // [true, Result] | exists X. [false, MLFunction<Result, X>, X]
        while (!result[0]) {
            const g = result[1];
            if ("_MLTAIL_" in g) {
                result = g._MLTAIL_.apply(undefined, result[2]);
            } else {
                return g.apply(undefined, result[2]);
            }
        }
        return result[1];
    }
    F._MLTAIL_ = f;
    return F;
}
//END
//BEGIN _function
function _function(f) {
    return function() {
        return f(arguments);
    };
}
//END

File: lib/lunarml/mlinit.lua
--BEGIN assert
local assert = assert
--END
--BEGIN error
local error = error
--END
--BEGIN getmetatable
local getmetatable = getmetatable
--END
--BEGIN ipairs
local ipairs = ipairs
--END
--BEGIN pairs
local pairs = pairs
--END
--BEGIN pcall
local pcall = pcall
--END
--BEGIN setmetatable
local setmetatable = setmetatable
--END
--BEGIN math
local math = math
--END
--BEGIN math_abs: math
local math_abs = math.abs
--END
--BEGIN math_type: math
local math_type = math.type
--END
--BEGIN math_maxinteger: math
local math_maxinteger = math.maxinteger
--END
--BEGIN math_mininteger: math
local math_mininteger = math.mininteger
--END
--BEGIN math_ult: math
local math_ult = math.ult
--END
--BEGIN string
local string = string
--END
--BEGIN string_char: string
local string_char = string.char
--END
--BEGIN string_format: string
local string_format = string.format
--END
--BEGIN table
local table = table
--END
--BEGIN table_concat: table
local table_concat = table.concat
--END
--BEGIN table_pack: table
local table_pack = table.pack
--END
--BEGIN table_unpack: table
local table_unpack = table.unpack
--END
--BEGIN table_unpackN: table_unpack
local function table_unpackN(t)
  return table_unpack(t, 1, t.n)
end
--END

--BEGIN _id
local function _id(x)
  return x
end
--END

--BEGIN _method
local function _method(obj, name, ...)
  return obj[name](obj, ...)
end
--END

--BEGIN _mkFn2
local function _mkFn2(f)
  return function(a, b)
    return f({a, b})
  end
end
--END
--BEGIN _mkFn3
local function _mkFn3(f)
  return function(a, b, c)
    return f({a, b, c})
  end
end
--END

--BEGIN _exn_meta: string_format
local _exn_meta = {}
function _exn_meta:__tostring()
  local traceback = self.traceback
  if traceback then
    traceback = "\n" .. traceback
  else
    traceback = ""
  end
  return string_format("%s: %s%s", self.location or "<no location info>", self.tag[1], traceback)
end
--END
--BEGIN _Match_tag
local _Match_tag = { "Match" }
--END
--BEGIN _Match: setmetatable _Match_tag _exn_meta
local _Match = setmetatable({ tag = _Match_tag }, _exn_meta)
--END
--BEGIN _isMatch: getmetatable _Match_tag _exn_meta
local function _isMatch(e)
  return getmetatable(e) == _exn_meta and e.tag == _Match_tag
end
--END
--BEGIN _Bind_tag
local _Bind_tag = { "Bind" }
--END
--BEGIN _Bind: setmetatable _Bind_tag _exn_meta
local _Bind = setmetatable({ tag = _Bind_tag }, _exn_meta)
--END
--BEGIN _isBind: getmetatable _Bind_tag _exn_meta
local function _isBind(e)
  return getmetatable(e) == _exn_meta and e.tag == _Bind_tag
end
--END
--BEGIN _Overflow_tag
local _Overflow_tag = { "Overflow" }
--END
--BEGIN _Overflow: setmetatable _Overflow_tag _exn_meta
local _Overflow = setmetatable({ tag = _Overflow_tag }, _exn_meta)
--END
--BEGIN _isOverflow: getmetatable _Overflow_tag _exn_meta
local function _isOverflow(e)
  return getmetatable(e) == _exn_meta and e.tag == _Overflow_tag
end
--END
--BEGIN _Div_tag
local _Div_tag = { "Div" }
--END
--BEGIN _Div: setmetatable _Div_tag _exn_meta
local _Div = setmetatable({ tag = _Div_tag }, _exn_meta)
--END
--BEGIN _isDiv: getmetatable _Div_tag _exn_meta
local function _isDiv(e)
  return getmetatable(e) == _exn_meta and e.tag == _Div_tag
end
--END
--BEGIN _Size_tag
local _Size_tag = { "Size" }
--END
--BEGIN _Size: setmetatable _Size_tag _exn_meta
local _Size = setmetatable({ tag = _Size_tag }, _exn_meta)
--END
--BEGIN _isSize: getmetatable _Size_tag _exn_meta
local function _isSize(e)
  return getmetatable(e) == _exn_meta and e.tag == _Size_tag
end
--END
--BEGIN _Subscript_tag
local _Subscript_tag = { "Subscript" }
--END
--BEGIN _Subscript: setmetatable _Subscript_tag _exn_meta
local _Subscript = setmetatable({ tag = _Subscript_tag }, _exn_meta)
--END
--BEGIN _isSubscript: getmetatable _Subscript_tag _exn_meta
local function _isSubscript(e)
  return getmetatable(e) == _exn_meta and e.tag == _Subscript_tag
end
--END
--BEGIN _Fail_tag
local _Fail_tag = { "Fail" }
--END
--BEGIN _Fail: setmetatable _Fail_tag _exn_meta
local function _Fail(message)
  return setmetatable({ tag = _Fail_tag, payload = message }, _exn_meta)
end
--END
--BEGIN _isFail: getmetatable _Fail_tag _exn_meta
local function _isFail(e)
  return getmetatable(e) == _exn_meta and e.tag == _Fail_tag
end
--END
--BEGIN _Fail_payload
local function _Fail_payload(e)
  return e.payload
end
--END
--BEGIN _isError: getmetatable _exn_meta
local function _isError(e)
  return getmetatable(e) ~= _exn_meta
end
--END

--BEGIN _exnName: getmetatable _exn_meta
local function _exnName(e)
  if getmetatable(e) == _exn_meta then
    return e.tag[1]
  else
    return "Error"
  end
end
--END

--BEGIN __exn_instanceof: getmetatable _exn_meta
local function __exn_instanceof(e, tag)
  return getmetatable(e) == _exn_meta and e.tag == tag
end
--END

--BEGIN _raise: getmetatable setmetatable _exn_meta error
local function _raise(x, location)
  local e
  if getmetatable(x) == _exn_meta and location ~= nil then
    local traceback = debug.traceback(nil, 2)
    e = setmetatable({ tag = x.tag, payload = x.payload, location = location, traceback = traceback }, _exn_meta)
  else
    e = x
  end
  error(e, 1)
end
--END

-- Int
--BEGIN _Int_add: _raise _Overflow
local function _Int_add(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  local z = x + y
  if y > 0 and z < x then
    _raise(_Overflow, "Int.+")
  elseif y < 0 and z > x then
    _raise(_Overflow, "Int.+")
  else
    return z
  end
end
--END
--BEGIN _Int_sub: _raise _Overflow
local function _Int_sub(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  local z = x - y
  if y < 0 and z < x then
    _raise(_Overflow, "Int.-")
  elseif y > 0 and x < z then
    _raise(_Overflow, "Int.-")
  else
    return z
  end
end
--END
--BEGIN _Int_mul: _raise _Overflow
local function _Int_mul(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  local z = x * y
  if (x ~= 0 and z // x ~= y) or (y ~= 0 and z // y ~= x) then
    _raise(_Overflow, "Int.*")
  else
    return z
  end
end
--END
--BEGIN _Int_div: _raise _Div math_mininteger _Overflow
local function _Int_div(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Int.div")
  elseif x == math_mininteger and y == -1 then
    _raise(_Overflow, "Int.div")
  end
  return x // y
end
--END
--BEGIN _Int_mod: _raise _Div
local function _Int_mod(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Int.mod")
  end
  return x % y
end
--END
--BEGIN _Int_negate: math_mininteger _raise _Overflow
local function _Int_negate(x)
  -- assert(math_type(x) == "integer")
  if x == math_mininteger then
    _raise(_Overflow, "Int.~")
  end
  return - x
end
--END
--BEGIN _Int_abs: math_mininteger _raise _Overflow math_abs
local function _Int_abs(x)
  -- assert(math_type(x) == "integer")
  if x == math_mininteger then
    _raise(_Overflow, "Int.abs")
  end
  return math_abs(x)
end
--END

-- Word
--BEGIN _Word_div: _raise _Div math_ult
local function _Word_div(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Word.div")
  elseif y > 0 then
    if x >= 0 then
      return x // y
    else -- x < 0
      -- Algorithm from Programming in Lua, 4th ed.
      local q = ((x >> 1) // y) << 1
      local r = x - q * y
      if math_ult(r, y) then
        return q
      else
        return q + 1
      end
    end
  else -- y < 0
    if math_ult(x, y) then
      return 0
    else
      return 1
    end
  end
end
--END
--BEGIN _Word_mod: _raise _Div math_ult
local function _Word_mod(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Word.mod")
  elseif y > 0 then
    if x >= 0 then
      return x % y
    else -- x < 0
      local q = ((x >> 1) // y) << 1
      local r = x - q * y
      if math_ult(r, y) then
        return r
      else
        return r - y
      end
    end
  else -- y < 0
    if math_ult(x, y) then
      return x
    else
      return x - y
    end
  end
end
--END

-- List
--BEGIN _list
local function _list(t)
  local xs = nil
  for i = t.n, 1, -1 do
    xs = { t[i], xs }
  end
  return xs
end
--END

-- Vector/Array
--BEGIN _Array_array: _raise _Size
local function _Array_array(n, init)
  if n < 0 then -- or maxLen < n
    _raise(_Size, "Array.array")
  end
  local t = { n = n }
  for i = 1, n do
    t[i] = init
  end
  return t
end
--END
--BEGIN _VectorOrArray_fromList
local function _VectorOrArray_fromList(xs)
  local t = {}
  local n = 0
  while xs ~= nil do
    n = n + 1
    t[n] = xs[1]
    xs = xs[2]
  end
  t.n = n
  return t
end
--END
--BEGIN _Vector_unsafeFromListRevN
local function _Vector_unsafeFromListRevN(n, xs)
  local t = { n = n }
  while xs ~= nil do
    t[n] = xs[1]
    xs = xs[2]
    n = n - 1
  end
  -- n must be 0
  return t
end
--END
--BEGIN _VectorOrArray_tabulate: _raise _Size
local function _VectorOrArray_tabulate(n, f)
  if n < 0 then -- or maxLen < n
    _raise(_Size, "(Vector|Array).tabulate")
  end
  local t = { n = n }
  for i = 1, n do
    t[i] = f(i - 1)
  end
  return t
end
--END

-- Vector
--BEGIN _Vector_concat
local function _Vector_concat(xs)
  local n = 0
  local t = {}
  while xs ~= nil do
    local u = xs[1]
    local m = u.n
    for i = 1,m do
      t[n + i] = u[i]
    end
    n = n + m
    xs = xs[2]
  end
  t.n = n
  return t
end
--END

-- Lua interface
--BEGIN _Lua_function: table_pack table_unpack
local function _Lua_function(f)
  return function(...)
    local r = f(table_pack(...))
    return table_unpack(r, 1, r.n)
  end
end
--END
--BEGIN _newTableWith: ipairs
local function _newTableWith(entries)
  local t = {}
  for _, p in ipairs(entries) do
    t[p[1]] = p[2]
  end
  return t
end
--END
